docs/bsescm.1.md
% BSESCM(1) Beast-@BUILDID@ | Beast/Bse Manual
%
% @FILE_REVISION@

# NAME
bsescm - A Guile based scheme shell for Beast and Bse

# SYNOPSIS

**bsescm** \[*options*\] \[**--**\] **…**

**bsescm** \[*options*\] *\<infile.bse\>* **…**

# DESCRIPTION

**Bsescm** is a **Guile**(1) based scheme shell for <a href="https://testbit.eu/wiki.mw/index.php?title=Bse&amp;action=edit&amp;redlink=1" class="new" title="Bse (page does not exist)">Bse</a>, the Beast Sound Engine.

Bsescm provides a shell interface to all procedures exported by the Bse library, so scripts are able to access the full range of functionality provided by it, from simple playback of a Bse file to full fledged automated creation or editing of synthesis networks.

Alternatively to the linked in Bse library, the Bsescm language interface can also talk to a remotely running Bse library host, for instance **Beast**(1). By operating on a remotely running Bse instance, Bsescm can be used to script arbitrary Bse programs. The shell is also used directly by the Bse library to execute procedures on its behalf, this allows Bse procedures to be written in scheme.

When started, Bsescm tests whether the first non-option argument is a Bse file and if the test succeeds, attempts to play the command line arguments as Bse files.

# OPTIONS

Bsescm follows the usual GNU command line syntax, with long options starting with two dashes ('--').

**--bse-pipe** *INFD* *OUTFD*
:   Provide the input and output communication filedescriptors for remote operation.

**--bse-eval** *STRING*
:   Execute (eval-string *STRING*) instead of going into interactive mode.

**--bse-enable-register**
:   Allows registration of procedures with the Bse library.

**--bse-no-load**
:   Prevent automated loading of plugins and scripts at startup time in interactive mode.

**--bse-no-play**
:   Prevent automated detection and playback of Bse file command line arguments.

**--g-fatal-warnings**
:   Make runtime warnings fatal (abort).

**-h**, **--help**
:   Describe command line options and exit.

**-v**, **--version**
:   Display version and exit.

**-n**=*NICELEVEL*
:   Execute with priority *NICELEVEL*, this option only takes effect for the root suid wrapper 'beast'.

**-N**
:   Disable renicing to execute with existing priority.

**--bse-latency**=*MSECONDS*
:   Set the allowed synthesis latency for Bse in milliseconds.

**--bse-mixing-freq**=*FREQUENCY*
:   Set the desired synthesis mixing frequency in Hz.

**--bse-control-freq**=*FREQUENCY*
:   Set the desired control frequency in Hz, this should be much smaller than the synthesis mixing frequency to reduce CPU load. The default value of approximately 1000 Hz is usually a good choice.

**--bse-pcm-driver** *DRIVERCONF*

**-p** *DRIVERCONF*
:   This options results in an attempt to open the PCM driver *DRIVERCONF* when playback is started. Multiple options may be supplied to try a variety of drivers and unless *DRIVERCONF* is specified as "**auto**", only the drivers listed by options are used. Each *DRIVERCONF* consists of a driver name and an optional comma seperated list of arguments attached to the driver withan equal sign, e.g.: **-p oss=/dev/dsp2,rw -p auto**.

**--bse-midi-driver** *DRIVERCONF*

**-m** *DRIVERCONF*
:   This option is similar to the **--bse-pcm-driver** option, but applies to MIDI drivers and devices. It also may be specified multiple times and features an "**auto**" driver.

**--bse-driver-list**
:   Produce a list of all available PCM and MIDI drivers and available devices.

### Guile Options:

**-l** *FILE*
:   Load scheme source code from *FILE*.

**-e** *FUNCTION*
:   After reading script, apply *FUNCTION* to command-line arguments.

**-ds**
:   Do -s *SCRIPT* at this point (note that this argument must be used in conjuction with -s).

**--debug**
:   Start with debugging evaluator and backtraces enabled (useful for debugging scripts).

**--emacs**
:   Enable emacs protocol for use from within emacs (experimental).

The remaining Guile options stop argument processing, and pass all remaining arguments as the value of (command-line):

**--**
:   Stop argument processing, start in interactive mode.

**-c** *EXPR*
:   Stop argument processing, evaluate *EXPR* as a scheme expression.

**-s** *SCRIPT*
:   Load Scheme source from *SCRIPT* and execute as a script.


# SEE ALSO

[**beast(1)**](beast.1.html),
[**bse(5)**](bse.5.html),
[**Beast/Bse Website**](http://beast.testbit.eu){.external},
[**Guile**](http://www.gnu.org/software/guile/docs/){.external}


NEWS
Beast 0.10.0:					(2016-04-07)

* Memory leak and stability fixes.
* Removal of deprecated Glib/Gtk+ functions.
* A new binary 'bsetool' subsumes several old utilities.
* Removal of unused utilities and several places with dead code.
* Several classes are now (fully) ported to new AIDA IDL: SNet Track
  TrackPart Bus SampleFileInfo Project MusicalTuning NoteDescription Icon
* More porting work to use Rapicorn C++11 utilities.
* Fixes for g++-4.9 which has become a hard dependency.
* Support and use the TAP protocol for unit test evaluation.
* New versioning: Releases increase MAJOR for major API changes.
* New versioning: Releases increase MINOR in case of any ABI breaks.
* New versioning: Releases increase MICRO for backwards compatibility.
* Now libbse supports parallel installation of runtime library packages.
* Now various program and module installation paths include MAJOR.
* Fix build issues with external libbse plugins.
* Fix LADSPAv1 plugin registration failing to detect some plugins.
* Provide C++11 API for undo/redo functions.
* Add pybse API wrapper that allows Python to 'import Bse'.
* Start a Python interpreter from Beast.
* Split up Beast main() methods to run event loop from Python.
* Port libbse to use C++11 scoped enums.

Overview of Changes in BEAST/BSE 0.9.2:		(2015-09-16)

* Workarounds were added for broken GLib compatibility.
* Files for the BSE runtime are now compressed and stored via rapidres(1).
* A good chunk of the Beast & Bse API has been migrated to use Rapicorn's AIDA IDL.
* The complexity of the BSE startup process could be reduced.
* A workaround for crashes in vorbisfile-1.3.4 and earlier was added.
* Shutdown crashes were fixed by stopping BSE threads via atexit.
* A new bse tool with commands check-load and render2wav replaces SCM scripts.
* Lots of fixes and optimizations for parallel builds were applied.
* All files are now installed under versioned directories, e.g. /usr/share/beast-0.9/
* Execute audio tests with 'make check' for development versions.

Overview of Changes in BEAST/BSE 0.9.0:		(2014-10-28)

* Fixed missing includes. [stw]
* Updates to build with Rapicorn1410.
* Work around GClosure API breakages in GLib.
* Work around GLib API deprecations.
* Remove old autotools and libtool workarounds.
* Updates to compile with g++-4.9.1.
* Fix issues on 32bit platforms.
* Optimize ELF relocations.
* Dependency update to Python 2.7.5.
* Updates to the manual pages.
* Improved utilization of compiler vectorizations.

Overview of Changes in BEAST/BSE 0.8.2:		(2013-07-06)

* Introduced Rapicorn1307 dependency.
* Establish Aida remote interfacing between Beast and BSE.
* Removed birnet/ sublibrary entirely, replaced with Rapicorn uses.
* Added support for loading and saving of FLAC encoded samples. [stw]
* Ported code to use Rapicorn's type safe cxxprintf instead of vprintf.
* Lots of documentation improvements and fixes.
* Automatization of documentation builds documentation uploads.
* Beast now ships complete documentation builds in tarballs.


Overview of Changes in BEAST/BSE 0.8.0:		(2013-03-25)

* Added use of the Rapicorn library, require version 13.0.0.
* Adapted IDL types to other projects.
* Adapted code to make use of C++11 and STL threads.
* Various performance improvements through EventFD, Spinlocks, lambdas, etc.
* Major documentation improvements, see: http://dev.testbit.eu/beast/latest/
* Fixed printf format-security issues (#665097). [Alessio Treglia]
* Lots of code cleanups, licensing cleanups.


Overview of Changes in BEAST/BSE 0.7.8:

* Wave oscillator fixes for 96k samples (bug #612281). [stw]
* Performance improvements of the scheme shell. [stw]
* Allow Alt+MouseButton1 to drag synthesis modules.
* Merged plugin libraries, improved build times.
* Use the ALSA 'default' PCM driver by default.
* Adapt code to allow fftw drop ins. [stw]
* Migrated code base to compile as C++11.
* Lots of simplifications and cleanups.
* Extended MIDI device listing.
* Improved Beast startup time.


Overview of Changes in BEAST/BSE 0.7.6:

* Packaging fixups, download and integrate documentation from beast wiki.
* Work around regressions introduced in recent GLib and GTK+ versions.
* Various bug and leak fixups.
* Moved lots of code to C++.
* Build documentation with doxygen.
* Code fixups for g++-4.4, g++-4.7.
* Improved DavOrgan frequency handling.


Overview of Changes in BEAST/BSE 0.7.4:

* Renamed the project to Better Audio System / Better Sound Engine
* Moved project website to: http://beast.testbit.eu/
* Various build system fixes [stw,timj]
* License fixups for some scripts [stw]
* Fixed subnormal tests on AMD64 if SSE unit is in DAZ mode [stw]
* Replaced slow resampler checks with a much faster resampling test [stw]
* Performance improvements for various tests [stw]
* GLib 2.28 unit test porting [stw]
* Speed improvements for record field name [stw]
* Fixed XRUNs in ALSA driver on 64bit systems [timj]
* Added beast.doap [Jonh Wendell]
* PO handling improvements.
* Updated German translation.
* Updated Norwegian bokmål translation [Kjartan Maraas]
* Added e-Telugu translation [Veeven]


Overview of Changes in BEAST/BSE 0.7.2:

* Moved Beast/BSE to GNU LGPL, use AS-IS license for examples
* Module changes and additions:
  ArtsCompressor       - Relicensed to LGPL with permission from Matthias Kretz
  BseContribSampleAndHold - Relicensed to LGPL with permission from Artem Popov
  DavXTalStrings          - Use deterministic random numbers for unit tests
  BseNoise                - Improved random number generator
* Switched to autogenerated ChangeLogs
* Error bell can be muted in beast preferences dialog
* Added multisample creation/editing command line tool: bsewavetool
* Support adjustable volume, pitching and drum envelopes in .bsewave files
* Added Retro Acoustic drum kit [Tim, Stefan]
* New loadable Instruments/Effects:
  - BQS Bass Drum E8012 [Tim, Stefan]
  - BQS Slow Hum [Stefan, William DeVore]

shell/bsescm.c
/* BSE-SCM - Better Sound Engine Scheme Wrapper
 * Copyright (C) 2002 Tim Janik
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * A copy of the GNU Lesser General Public License should ship along
 * with this library; if not, see http://www.gnu.org/copyleft/.
 */
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <guile/gh.h>
#include <bse/bse.h>
#include <bse/bsemain.h>  /* for bse_init_textdomain_only() */
#include <sfi/sfistore.h> /* no bin-compat */
#include <sys/time.h>
#include <libintl.h>
#include <sys/resource.h>
#include "bsescminterp.h"
#include "topconfig.h"

#define	PRG_NAME	"bsescm"

#define BSE_EXIT_STATUS 102


/* --- prototypes --- */
static void	gh_main			(gint	 argc,
					 gchar	*argv[]);
static void	shell_parse_args	(gint    *argc_p,
					 gchar ***argv_p);
static void     shell_print_usage       (void);


/* --- variables --- */
static gint            bse_scm_pipe[2] = { -1, -1 };
static gchar          *bse_scm_eval_expr = NULL;
static gboolean        bse_scm_enable_register = FALSE;
static gboolean        bse_scm_auto_load = TRUE;
static gboolean        bse_scm_auto_play = TRUE;
static SfiComPort     *bse_scm_port = NULL;
static SfiGlueContext *bse_scm_context = NULL;
static const gchar    *boot_script_path = BSE_PATH_SCRIPTS;

/* --- functions --- */
static void
port_closed (SfiComPort *port,
	     gpointer    close_data)
{
  /* we don't do anything fancy here */
  if (port)
    exit (BSE_EXIT_STATUS);
}

int
main (int   argc,
      char *argv[])
{
  const gchar *env_str;

  sfi_init (&argc, &argv, "BSESCM", NULL);
  bse_init_textdomain_only();
  setlocale (LC_ALL, "");

  env_str = g_getenv ("BSESCM_SLEEP4GDB");
  if (env_str && atoi (env_str) >= 3)
    {
      g_message ("going into sleep mode due to debugging request (pid=%u)", getpid ());
      g_usleep (2147483647);
    }

  shell_parse_args (&argc, &argv);

  if (env_str && (atoi (env_str) >= 2 ||
                  (atoi (env_str) >= 1 && !bse_scm_enable_register)))
    {
      g_message ("going into sleep mode due to debugging request (pid=%u)", getpid ());
      g_usleep (2147483647);
    }

  if (bse_scm_pipe[0] >= 0 && bse_scm_pipe[1] >= 0)
    {
      bse_scm_port = sfi_com_port_from_pipe (PRG_NAME, bse_scm_pipe[0], bse_scm_pipe[1]);
      sfi_com_port_set_close_func (bse_scm_port, port_closed, NULL);
      if (!bse_scm_port->connected)
	{
	  g_printerr ("%s: failed to connect to pipe (%d, %d)\n", PRG_NAME, bse_scm_pipe[0], bse_scm_pipe[1]);
	  exit (BSE_EXIT_STATUS);
	}
      bse_scm_context = sfi_glue_encoder_context (bse_scm_port);
    }

  if (!bse_scm_context)
    {
      /* start our own core thread */
      bse_init_async (&argc, &argv, "BSESCM", NULL);
      bse_scm_context = bse_init_glue_context (PRG_NAME);
    }

  /* now that the BSE thread runs, drop scheduling priorities if we have any */
  setpriority (PRIO_PROCESS, getpid(), 0);

  gh_enter (argc, argv, gh_main);

  return 0;
}


static void
gh_main (int   argc,
	 char *argv[])
{
  /* initial interpreter setup */
  if (bse_scm_enable_register)
    bse_scm_enable_script_register (TRUE);
  else
    bse_scm_enable_server (TRUE);
  sfi_glue_context_push (bse_scm_context);

  /* initialize interpreter */
  bse_scm_interp_init ();

  /* exec Bse Scheme bootup code */
  const gchar *boot_script = g_intern_printf ("%s/%s", boot_script_path, "bse-scm-glue.boot");
  gh_load (boot_script);

  /* eval, auto-play or interactive */
  if (bse_scm_eval_expr)
    gh_eval_str (bse_scm_eval_expr);
  else
    {
      gboolean call_auto_play = FALSE;
      /* detect bse file */
      if (argc >= 2 && bse_scm_auto_play)
        {
          SfiRStore *rstore = sfi_rstore_new_open (argv[1]);
          if (rstore)
            {
              if (g_scanner_get_next_token (rstore->scanner) == '(' &&
                  g_scanner_get_next_token (rstore->scanner) == G_TOKEN_IDENTIFIER)
                {
                  if (strcmp ("bse-version", rstore->scanner->value.v_string) == 0 &&
                      g_scanner_get_next_token (rstore->scanner) == G_TOKEN_STRING &&
                      g_scanner_get_next_token (rstore->scanner) == ')')
                    call_auto_play = TRUE;
                }
              sfi_rstore_destroy (rstore);
            }
        }
      /* auto-play or interactive */
      if (call_auto_play)
        gh_eval_str ("(bse-scm-auto-play)");
      else 
        {
          if (bse_scm_auto_load)
            gh_eval_str ("(bse-server-register-blocking bse-server-register-core-plugins #f)"
                         "(bse-server-register-blocking bse-server-register-scripts #f)"
                         "(bse-server-register-blocking bse-server-register-ladspa-plugins #f)");
          gh_repl (argc, argv);
        }
    }

  /* shutdown */
  sfi_glue_context_pop ();
  if (bse_scm_port)
    {
      sfi_com_port_set_close_func (bse_scm_port, NULL, NULL);
      sfi_com_port_close_remote (bse_scm_port, FALSE);
    }
  sfi_glue_context_destroy (bse_scm_context);
}

static void
shell_parse_args (gint    *argc_p,
		  gchar ***argv_p)
{
  guint argc = *argc_p;
  gchar **argv = *argv_p;
  guint i, e;
  gboolean initialize_bse_and_exit = FALSE;

  for (i = 1; i < argc; i++)
    {
      if (strcmp (argv[i], "--") == 0 ||
          strcmp (argv[i], "-s") == 0 ||
          strcmp (argv[i], "-c") == 0)
	break;
      else if (strcmp (argv[i], "--g-fatal-warnings") == 0)
	{
	  GLogLevelFlags fatal_mask;

	  fatal_mask = g_log_set_always_fatal (G_LOG_FATAL_MASK);
	  fatal_mask |= G_LOG_LEVEL_WARNING | G_LOG_LEVEL_CRITICAL;
	  g_log_set_always_fatal (fatal_mask);
	  argv[i] = NULL;
	}
      else if (strcmp (argv[i], "--bse-pipe") == 0)
	{
	  /* first two arguments are input and output pipes from BseServer */
	  bse_scm_pipe[0] = -1;
	  bse_scm_pipe[1] = -1;
	  if (i + 2 < argc)
	    {
	      bse_scm_pipe[0] = atoi (argv[i + 1]);
	      bse_scm_pipe[1] = atoi (argv[i + 2]);
	      argv[i++] = NULL;
	      argv[i++] = NULL;
	    }
	  argv[i] = NULL;
	  if (bse_scm_pipe[0] < 2 || bse_scm_pipe[1] < 2)
	    {
	      g_printerr ("%s: invalid arguments supplied for: --bse-pipe <inpipe> <outpipe>\n", PRG_NAME);
	      exit (BSE_EXIT_STATUS);
	    }
	}
      else if (strcmp (argv[i], "--bse-eval") == 0)
	{
	  bse_scm_eval_expr = NULL;
	  if (i + 1 < argc)
	    {
	      bse_scm_eval_expr = argv[i + 1];
	      argv[i++] = NULL;
	    }
	  argv[i] = NULL;
	  if (!bse_scm_eval_expr)
	    {
	      g_printerr ("%s: invalid arguments supplied for: --bse-eval <expression>\n", PRG_NAME);
	      exit (BSE_EXIT_STATUS);
	    }
	}
      else if (strcmp (argv[i], "--bse-enable-register") == 0)
	{
	  bse_scm_enable_register = TRUE;
	  argv[i] = NULL;
	}
      else if (strcmp (argv[i], "--bse-no-load") == 0)
	{
	  bse_scm_auto_load = FALSE;
	  argv[i] = NULL;
	}
      else if (strcmp (argv[i], "--bse-no-play") == 0)
	{
	  bse_scm_auto_play = FALSE;
	  argv[i] = NULL;
	}
      else if (strcmp ("--bse-driver-list", argv[i]) == 0)
        {
          initialize_bse_and_exit = TRUE;
          /* leave args for BSE */
        }
      else if (strcmp ("-p", argv[i]) == 0)
        {
          /* modify args for BSE */
          argv[i] = "--bse-pcm-driver";
        }
      else if (strcmp ("-m", argv[i]) == 0)
        {
          /* modify args for BSE */
          argv[i] = "--bse-midi-driver";
        }
      else if (strcmp ("--bse-override-script-path", argv[i]) == 0 && i + 1 < argc)
        {
          argv[i++] = NULL;
          boot_script_path = argv[i];
          argv[i] = NULL;
        }
      else if (strcmp (argv[i], "-n") == 0 && i + 1 < argc)
        { /* handled by priviledged launcher */
          argv[i++] = NULL;
          argv[i] = NULL;
        }
      else if (strncmp (argv[i], "-n=", 3) == 0 ||
               (strncmp (argv[i], "-n", 2) == 0 && argv[i][3] >= '0' && argv[i][3] <= '9') ||
               strcmp (argv[i], "-N") == 0)
        { /* handled by priviledged launcher */
          argv[i] = NULL;
        }
      else if (strcmp ("-h", argv[i]) == 0 ||
               strcmp ("--help", argv[i]) == 0)
        {
          shell_print_usage();
          exit (0);
        }
      else if (strcmp ("-v", argv[i]) == 0 ||
               strcmp ("--version", argv[i]) == 0)
        {
          g_print ("BSESCM version %s (%s)\n", BST_VERSION, BST_VERSION_HINT);
          g_print ("Libraries: ");
          g_print ("GLib %u.%u.%u", glib_major_version, glib_minor_version, glib_micro_version);
          g_print (", SFI %u.%u.%u", bse_major_version, bse_minor_version, bse_micro_version);
          g_print (", BSE %u.%u.%u", bse_major_version, bse_minor_version, bse_micro_version);
          g_print (", Guile %u.%u.%u", SCM_MAJOR_VERSION, SCM_MINOR_VERSION, SCM_MICRO_VERSION);
          g_print ("\n");
          g_print ("Compiled for: %s\n", BST_ARCH_NAME);
          g_print ("BSESCM comes with ABSOLUTELY NO WARRANTY.\n");
          g_print ("You may redistribute copies of BSESCM under the terms of\n");
          g_print ("the GNU Lesser General Public License which can be found in\n");
          g_print ("the BEAST source package. Sources, examples and contact\n");
          g_print ("information are available at http://beast.testbit.eu/.\n");
          exit (0);
        }
    }
  
  e = 1;
  for (i = 1; i < argc; i++)
    if (argv[i])
      {
        argv[e++] = argv[i];
        if (i >= e)
          argv[i] = NULL;
      }
  *argc_p = e;

  if (initialize_bse_and_exit)
    {
      bse_init_async (argc_p, argv_p, "BSESCM", NULL);
      exit (0);
    }
}

static void
shell_print_usage (void)
{
  g_print ("Usage: bsescm [options] [files...]\n");
  g_print ("Play BSE files and evaluate Scheme code, interactively or from a script.\n");
  g_print ("  -h, --help              show this help message\n");
  g_print ("  -v, --version           print version and exit\n");
  g_print ("  -n NICELEVEL            run with priority NICELEVEL (for suid wrapper bsescm)\n");
  g_print ("  -N                      disable renicing\n");
  g_print ("  --bse-pipe INFD OUTFD   remote operation filedescriptors\n");
  g_print ("  --bse-eval STRING       execute (eval-string STRING) and exit\n");
  g_print ("  --bse-enable-register   allow registration of BSE procedures\n");
  g_print ("  --bse-no-load           prevent automated plugin and script registration\n");
  g_print ("  --bse-no-play           prevent detection of BSE file command line arguments\n");
  g_print ("  --bse-latency=USECONDS  specify synthesis latency in milliseconds\n");
  g_print ("  --bse-mixing-freq=FREQ  specify synthesis mixing frequency in Hz \n");
  g_print ("  --bse-control-freq=FREQ specify control frequency in Hz\n");
  g_print ("  --bse-pcm-driver DRIVERCONF\n");
  g_print ("  -p DRIVERCONF           try to use the PCM driver DRIVERCONF, multiple\n");
  g_print ("                          options may be supplied to try a variety of\n");
  g_print ("                          drivers, unless -p auto is given, only the\n");
  g_print ("                          drivers listed by -p options are used; each\n");
  g_print ("                          DRIVERCONF consists of a driver name and an\n");
  g_print ("                          optional comma seperated list of arguments,\n");
  g_print ("                          e.g.: -p oss=/dev/dsp2,rw\n");
  g_print ("  --bse-midi-driver DRIVERCONF\n");
  g_print ("  -m DRIVERCONF           try to use the MIDI driver DRIVERCONF, multiple\n");
  g_print ("                          options may be specified similarly to the\n");
  g_print ("                          option handling for --bse-pcm-driver\n");
  g_print ("  --bse-driver-list       list available PCM and MIDI drivers\n");
  g_print ("Guile Options:\n");
  g_print ("  -s SCRIPT               load Scheme source code from FILE, and exit\n");
  g_print ("  -c EXPR                 evalute Scheme expression EXPR, and exit\n");
  g_print ("  --                      stop scanning arguments; run interactively\n");
  g_print ("The above switches stop argument processing, and pass all\n");
  g_print ("remaining arguments as the value of (command-line).\n");
  g_print ("  -l FILE                 load Scheme source code from FILE\n");
  g_print ("  -e FUNCTION             after reading script, apply FUNCTION to\n");
  g_print ("                          command line arguments\n");
  g_print ("  -ds                     do -s script at this point\n");
  g_print ("  --debug                 start with debugging evaluator and backtraces\n");
  g_print ("  -q                      inhibit loading of user init file\n");
  g_print ("  --emacs                 enable Emacs protocol (experimental)\n");
  g_print ("  --use-srfi=LS           load SRFI modules for the SRFIs in LS,\n");
  g_print ("                          which is a list of numbers like \"2,13,14\"\n");
  g_print ("  \\                       read arguments from following script lines\n");
}
